### PR TITLE
[dynamo][moco] Save global torch state to restore on graph break

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101201

This is relevant to  https://github.com/pytorch/pytorch/pull/100570 as well.

cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_guards.py [hunk 1]
```diff
@@ -344,6 +344,52 @@ def restore_graphstate(self, state):
...
+    For example, torch.is_grad_enabled.
+    """
+
+    def __init__(self):
+        self.global_state: Dict[str, bool] = {}
+
+    def copy_graphstate(self):
+        return GlobalContextCheckpointState(dict(self.global_state))
+
+    def restore_graphstate(self, state):
```
Comment (voznesenskym) [inline torch/_guards.py:382]: These need to be added to the copy/restore flow on TX

**Comment by voznesenskym** [comment]
Can you reference https://github.com/pytorch/pytorch/pull/100570 in your description? A lot of relevant conversation there.

File: torch/_guards.py (hunk 1 above)
Comment (anijain2305) [reply torch/_guards.py:382]: Can you elaborate more? I am calling these during the copy/restore in the `output_graph`

Comment (voznesenskym) [reply torch/_guards.py:382]: Oh I missed that this is in output graph. That works for me! 

Consider moving it up to Tx, it fits there better, but no strong opinions.

**Review by voznesenskym — APPROVED** [review]

File: torch/_dynamo/output_graph.py [hunk 2]
```diff
@@ -354,6 +362,10 @@ def guards(self) -> Set[Guard]:
     def nn_modules(self) -> Dict[str, torch.nn.Module]:
         return self.tracing_context.module_context.nn_modules
 
+    def save_global_state(self):
+        global_state = self.tracing_context.global_context.global_state
+        global_state["is_grad_enabled"] = torch.is_grad_enabled()
```
Comment (bdhirsh) [inline torch/_dynamo/output_graph.py:367]: basic question: there's other global state, like e.g. current autocast context. Should this function be checkpointing that too?

Comment (anijain2305) [reply torch/_dynamo/output_graph.py:367]: Yes, thats what I was thinking too. Yes, we should. Thats why I created this central place to handle global states. We will add more cases here as they come.

**Pull request closed without merging**